### ISSUE TITLE
Incorrect gradient of model output with respect to inputs using GradientTape

### ISSUE THREAD

**Issue by bzdjordje** [tensorflow/tensorflow#56341]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.1.3

### Custom Code

No

### OS Platform and Distribution

Windows 10 Home

### Python version

3.7

### Bazel version

n/a

### GCC/Compiler version

n/a

### CUDA/cuDNN version

CPU

### GPU model and memory

n/a

### Current Behaviour?

```shell
I am trying to take the gradient/jacobian of my model output with respect to the inputs to study the derivatives of my parameter space. This issue has come up several times in the past but I do not seem to be able to reproduce their results or solutions, if any were suggested. It seems that I am able to apply gradients a single variable output, single variable input model, however my problem is multivariate with respect to both. I am using an older version of TF for work purposes but the issue was the same for TF 2.6 as well.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
from numpy import *
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

x1 = linspace(0,4*pi,100000)
x2 = linspace(2,20,100000)
y = cos(x1) + x2/5

model = Sequential()
model.add(Dense(100, activation='elu', input_shape=(2,)))  # elu used so that smooth, continuous functions are differentiated
model.add(Dense(100, activation='elu'))
model.add(Dense(50, activation='elu'))
model.add(Dense(50, activation='elu'))
model.add(Dense(10, activation='elu'))
model.add(Dense(10, activation='elu'))
model.add(Dense(1, activation='linear'))
model.compile(loss='MSE', optimizer='adam', metrics=['accuracy'])

x_in = vstack((x1,x2)).T
y_out = y

x_train, x_valid, y_train, y_valid = train_test_split(x_in, y_out, test_size=0.3, shuffle= True)

epochs = 100
batch_size = 110

h0 = model.fit(x_train, y_train, validation_data=(x_valid,y_valid), batch_size=batch_size,epochs=epochs)

grado1 = zeros((100000,2))
grado2 = zeros((100000,2))

for i in range(0,100000):
    x = tf.Variable([[x_in[i,0],x_in[i,1]]])
    
    with tf.GradientTape(persistent=True,watch_accessed_variables=True) as t:
        t.watch(x)  
        out1 = model(x)
        out2 = tf.math.cos(x[:,0]) + x[:,1]/5  # this works using both a Variable array as well as separated taped variables, i.e., x1t, x2t
    gradients1 = t.gradient(out1, x)
    gradients2 = t.gradient(out2, x)
    if i % 100 == 0:
        print(i)
    grado1[i,:] = gradients1.numpy()
    grado2[i,:] = gradients2.numpy()


ymodel = model.predict([x_in])

for i in range(2):
    plt.figure(figsize=(15,4))
    ax = plt.subplot(111)
    ax.plot(x_in[:,i],y_out,label='data',linewidth=5)
    ax.plot(x_in[:,i],ymodel,'--',label='model',linewidth=5)
    plt.ylabel('y')
    plt.xlabel('x'+str(i+1))
    plt.legend(loc='lower left')
    
    ax2 = ax.twinx()
    ax2.plot(x_in[:,i],grado1[:,i],'b--',label='grad1_x'+str(i+1)+' model')
    ax2.plot(x_in[:,i],grado2[:,i],'g',label='grad2_x'+str(i+1)+' autodiff formula')
    if i == 0:
        ax2.plot(x_in[:,0],-sin(x_in[:,0]),'r--',label='grad2_x'+str(i+1)+' explicit formula')
    elif i == 1:
        ax2.plot(x_in[:,1],1/5*ones(len(x_in[:,1])),'r--',label='grad2_x'+str(i+1)+' explicit formula')
    plt.ylabel('dy/dx'+str(i+1))
    plt.xlabel('x'+str(i+1))

    plt.legend()
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@bzdjordje,
TensorFlow 2.1 is not actively supported. Could you update TensorFlow to the latest stable version v2.9 and check if you are facing the same issue. Thanks!

**Comment by bzdjordje** [comment]
Unfortunately the error still persists if I run the same script in v2.9 as presented below
![image](https://user-images.githubusercontent.com/4984500/171958551-8197cc91-2f19-4031-aaac-cc99482c5384.png)
![image](https://user-images.githubusercontent.com/4984500/171958562-3e646af7-1bee-4bd9-ad02-e2ab9d128031.png)

**Comment by hrbigelow** [comment]
Hi @tilakrayal and @bzdjordje,

So, I don't think this is a bug because you've trained your model in a very underdetermined fashion.  I think the better way would be to train it on a grid of points in $\mathbb{R}^2$ (that is, the cartesian product over ranges for x1 and x2.  Instead, you've trained it on a a single line `vstack(x1, x2)` where `x1 = linspace(0, 4*pi, 10000)` and `x2 = linspace(2, 20, 100000)`.

If you look at dy/dx2, the positions where it is very close to 5 (like around x2 = 18), the dotted line in the first graph is very close to that of the explicit formula.  And, places where dy/dx2 is very far from 5, like at x2=13,  then dy/dx1 is far from that of the analytic formula.  They are just compensating for each other.

**Comment by tilakrayal** [comment]
@bzdjordje,
Could you please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/56341#issuecomment-1146930162) from the user and let us know if it still an issue. Thank you!

**Comment by bzdjordje** [comment]
@tilakrayal @hrbigelow 
Thank you for the responses. The data is randomly shuffled prior to training so the model is not trained one slice as opposed to another, validation/training datasets are fairly uniformly distributed across.

zoomed in
![image](https://user-images.githubusercontent.com/4984500/172234304-919458e2-ba14-41d0-ab8b-a407f6b7c764.png)

(plotting only every 100th point)
![image](https://user-images.githubusercontent.com/4984500/172234443-009c93dd-a87e-4606-abed-ffa679af5424.png)

**Comment by hrbigelow** [comment]
Hi @bzdjordje,

When I said 'slice', I mean a subset of $\mathbb{R}^2$, which is the domain of your target function.  Actually, I mistook the domain.  You are taking 100000 points evenly spaced along the line segment from point (0, 2) to (4pi, 20).  What you should be calculating is the gradient projected to that line.  If you do that, you see that the autodiff gradient according to the model very nearly matches the analytical gradient of the function.

To calculate this gradient, you take the dot product with a normalized vector in the direction of that line.  Here is the result:

![image](https://user-images.githubusercontent.com/5355286/172240927-b929e454-f603-459f-a61f-80d6a7d5a555.png)

What I did is edited your code as follows:

```python
import tensorflow as tf
import numpy as np
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

def make_model():
    x1 = linspace(0,4*pi,100000)
    x2 = linspace(2,20,100000)
    y = cos(x1) + x2/5

    model = Sequential()
    model.add(Dense(100, activation='elu', input_shape=(2,)))  # elu used so that smooth, continuous functions are differentiated
    model.add(Dense(100, activation='elu'))
    model.add(Dense(50, activation='elu'))
    model.add(Dense(50, activation='elu'))
    model.add(Dense(10, activation='elu'))
    model.add(Dense(10, activation='elu'))
    model.add(Dense(1, activation='linear'))
    model.compile(loss='MSE', optimizer='adam', metrics=['accuracy'])

    x_in = np.vstack((x1,x2)).T
    y_out = y

    x_train, x_valid, y_train, y_valid = train_test_split(x_in, y_out, test_size=0.3, shuffle= True)

    epochs = 100
    batch_size = 110

    h0 = model.fit(x_train, y_train, validation_data=(x_valid,y_valid), batch_size=batch_size,epochs=epochs)

    return x_in, y_out, model

def analyze(model, x_in): 

    grado1 = np.zeros((100000,2))
    grado2 = np.zeros((100000,2))

    for i in range(0,100000):
        x = tf.Variable([[x_in[i,0],x_in[i,1]]])

        with tf.GradientTape(persistent=True,watch_accessed_variables=True) as t:
            t.watch(x)  
            out1 = model(x)
            out2 = tf.math.cos(x[:,0]) + x[:,1]/5  # this works using both a Variable array as well as separated taped variables, i.e., x1t, x2t
        gradients1 = t.gradient(out1, x)
        gradients2 = t.gradient(out2, x)
        if i % 100 == 0:
            print(i)
        grado1[i,:] = gradients1.numpy()
        grado2[i,:] = gradients2.numpy()

    ymodel = model.predict([x_in])

    return grado1, grado2, ymodel


print('x_in, y_out, model = make_model()')
print('grado1, grado2, ymodel = analyze(model, x_in)')

def main(x_in, y_out, grado1, grado2, ymodel):

    # project the gradients onto the line x1 = x2
    direction = np.array([4 * pi, 20 - 2])
    direction /= np.linalg.norm(direction)

    proj_grado1 = np.matmul(grado1, direction)
    proj_grado2 = np.matmul(grado2, direction)

    plt.figure(figsize=(15,4))
    ax = plt.subplot(111)
    plt.ylabel('dy / d_direction')
    plt.xlabel('x_1')
    plt.legend(loc='lower left')

    # ax2 = ax.twinx()
    # ax2.plot(x_in[:,i],grado1[:,i],'b--',label='grad1_x'+str(i+1)+' model')
    # ax2.plot(x_in[:,i],grado2[:,i],'g',label='grad2_x'+str(i+1)+' autodiff formula')

    ax.plot(x_in[:,0],proj_grado1[:],'b--',label='line_grad1_x model')
    ax.plot(x_in[:,0],proj_grado2[:],'g',label='line_grad2_x autodiff formula')
    # if i == 0:
        # ax2.plot(x_in[:,0],-sin(x_in[:,0]),'r--',label='grad2_x'+str(i+1)+' explicit formula')
    # elif i == 1:
        # ax2.plot(x_in[:,1],(1/5)*ones(len(x_in[:,1])),'r--',label='grad2_x'+str(i+1)+' explicit formula')
    # plt.ylabel('dy/dx_1')
    # plt.xlabel('x_1')

    plt.legend()
    plt.show()

```

**Comment by tilakrayal** [comment]
@bzdjordje,
Could you try to execute the above alternative work-around and confirm if it is working as expected. Thank you!

**Comment by bzdjordje** [comment]
@tilakrayal @hrbigelow 

Hello, the correction to my script provided by Henry worked on my computer, thank you!

**Comment by tilakrayal** [comment]
@bzdjordje,
 Glad the issue is resolved for you, please feel free to move this to closed status. Thank you!
